pkg/client/cache/deltastore.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+)
+
+// DeltaStore is a Store implementation that provides a sequence of compressed events to a consumer
+// along with event types.  This differs from the FIFO implementation in that FIFO does not provide
+// events when an object is deleted and does not provide the type of event.  Events are compressed
+// in a manner similar to FIFO, but accounting for event types and deletions.  The exact
+// compression semantics are as follows:
+//
+// 1.  If a watch.Added is enqueued with state X and a watch.Modified with state Y is received,
+//     these are compressed into (Added, Y)
+//
+// 2.  If a watch.Added is enqueued with state X and a watch.Deleted is received, these are
+//     compressed and the item is removed from the queue
+//
+// 3.  If a watch.Modified is enqueued with state X and a watch.Modified with state Y is received,
+//     these two events are compressed into (Modified, Y)
+//
+// 4.  If a watch.Modified is enqueued with state X and a watch.Deleted is received, these are
+//     compressed into (Deleted, X)
```

Comment:
Tricky - if the UID is different, I don't think it's a modified (for a lot of objects).

----- Original Message -----
> > +// along with event types.  This differs from the FIFO implementation in
> > that FIFO does not provide
> > +// events when an object is deleted and does not provide the type of
> > event.  Events are compressed
> > +// in a manner similar to FIFO, but accounting for event types and
> > deletions.  The exact
> > +// compression semantics are as follows:
> > +//
> > +// 1.  If a watch.Added is enqueued with state X and a watch.Modified with
> > state Y is received,
> > +//     these are compressed into (Added, Y)
> > +//
> > +// 2.  If a watch.Added is enqueued with state X and a watch.Deleted is
> > received, these are
> > +//     compressed and the item is removed from the queue
> > +//
> > +// 3.  If a watch.Modified is enqueued with state X and a watch.Modified
> > with state Y is received,
> > +//     these two events are compressed into (Modified, Y)
> > +//
> > +// 4.  If a watch.Modified is enqueued with state X and a watch.Deleted is
> > received, these are
> > +//     compressed into (Deleted, X)
> 
> What if a Watch.Delete(X) is enqueued and a Watch.Added(Y) is received?
> Compress to (Modified, Y)?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3521/files#r23730636